example/example-knex-mysql.js

```diff
@@ -0,0 +1,141 @@
+'use strict'
+
+// Example of a custom store using Knex.js and MySQL.
+//
+// Assumes you have access to a configured knex object.
+//
+// Note that the rate check should place a read lock on the row.
+// For MySQL see:
+// https://dev.mysql.com/doc/refman/8.0/en/innodb-locking-reads.html
+// https://blog.nodeswat.com/concurrency-mysql-and-node-js-a-journey-of-discovery-31281e53572e
+//
+// Below is an example table to store rate limits that must be created
+// in the database first.
+//
+// exports.up = async knex => {
+//   await knex.schema.createTable('rate_limits', table => {
+//     table.string('source').notNullable()
+//     table.string('route').notNullable()
+//     table.integer('count').unsigned()
+//     table.bigInteger ('ttl')
+//     table.primary(['route', 'source'])
+//   })
+// }
+//
+// exports.down = async knex => {
+//   await knex.schema.dropTable('rate_limits')
+// }
+//
+// CREATE TABLE `rate_limits` (
+//   `source` varchar(255) NOT NULL,
+//   `route` varchar(255) NOT NULL,
+//   `count` int unsigned DEFAULT NULL,
+//   `ttl` int unsigned DEFAULT NULL,
+//   PRIMARY KEY (`route`,`source`)
+// ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;
+
+function KnexStore(options) {
+  this.options = options
+  this.route = ''
+}
+
+KnexStore.prototype.routeKey = function (route) {
+  if (route) this.route = route
+  return route
+}
+
+KnexStore.prototype.incr = async function (key, cb) {
+  const now = (new Date()).getTime()
+  const ttl = now + this.options.timeWindow
+  const max = this.options.max
+  const cond = { route: this.route, source: key }
+  await knex.transaction(async (trx) => {
```

Comment:
Can we avoid mixing async/await with a callback?

Reply:
I think yes.
```js
let trx
try {
// Using trx as a transaction object:
  trx = await knex.transaction();
  await trx('') ....
  await trx('') ....
  await trx.commit()
} catch {
  await trx.rollback()
}
```
```

Reply:
Okay let me try. It may be a few days. Deadline looms. More soon...

Reply:
Thanks for the feedback. I've refactored as suggested.